### PR TITLE
Consolidate various BLOCKED_WAIT* states

### LINKED ISSUE #427: [NEW] Consolidate wait states
> I agree. I'm saying we already have the notion that a client is blocked with a pending command, that can get re-executed once the blocking is done. (That is how the normal command blocking works). The blocking mechanic is functionally the same as the other wait (wait for a replack from k replicas then unblock).

_Originally posted by @madolson in https://github.com/valkey-io/valkey/pull/245#discussion_r1588610039_

### PR DESCRIPTION
**Pull request by PingXie**
There are currently three block types: BLOCKED_WAIT, BLOCKED_WAITAOF, and BLOCKED_WAIT_PREREPL, used to block clients executing `WAIT`, `WAITAOF`, and `CLUSTER SETSLOT`, respectively. They share the same workflow: the client is blocked until replication to the expected number of replicas completes. However, they provide different responses depending on the commands involved. Using distinct block types leads to code duplication and reduced readability. This PR consolidates the three types into a single WAIT type, differentiating them using the pending command to ensure the appropriate response is returned.

Fix #427

### CODE REVIEW

**Comment by zuiderkwast** [comment]
Nice. Please add some useful text in the description, rather than just a link to the issue. We always ask that from contributors. It's helpful when reviewing, it's used as commit message when merging and it's good for reference later, even if the PR is not material for release notes.

**Comment by PingXie** [comment]
Will do

File: src/blocked.c [hunk 1]
```diff
@@ -231,15 +229,19 @@ void replyToBlockedClientTimedOut(client *c) {
         addReplyNullArray(c);
         updateStatsOnUnblock(c, 0, 0, 0);
     } else if (c->bstate.btype == BLOCKED_WAIT) {
-        addReplyLongLong(c, replicationCountAcksByOffset(c->bstate.reploffset));
-    } else if (c->bstate.btype == BLOCKED_WAITAOF) {
-        addReplyArrayLen(c, 2);
-        addReplyLongLong(c, server.fsynced_reploff >= c->bstate.reploffset);
-        addReplyLongLong(c, replicationCountAOFAcksByOffset(c->bstate.reploffset));
+        if (c->cmd == shared.wait_cmd) {
```
Comment (zuiderkwast) [inline src/blocked.c]: This is how the check for a particular command is done in many places:

```suggestion
        if (c->cmd->proc == waitCommad) {
```

No need for the shared objects.

File: src/blocked.c [hunk 2]
```diff
@@ -231,15 +229,19 @@ void replyToBlockedClientTimedOut(client *c) {
...
+            addReplyLongLong(c, replicationCountAcksByOffset(c->bstate.reploffset));
+        } else if (c->cmd == shared.waitaof_cmd) {
```
Comment (zuiderkwast) [inline src/blocked.c]:
```suggestion
        } else if (c->cmd->proc == waitaofCommand) {
```

File: src/blocked.c [hunk 3]
```diff
@@ -231,15 +229,19 @@ void replyToBlockedClientTimedOut(client *c) {
...
+            addReplyArrayLen(c, 2);
+            addReplyLongLong(c, server.fsynced_reploff >= c->bstate.reploffset);
+            addReplyLongLong(c, replicationCountAOFAcksByOffset(c->bstate.reploffset));
+        } else if (c->cmd == shared.setslot_cmd) {
```
Comment (zuiderkwast) [inline src/blocked.c]: CLUSTER SETSLOT is handled by clusterCommand, just like all cluster subcommands. We can either break it out to its own function clusterSetslotCommand and update the JSON file, or we can just check for clusterCommand here. It's the only cluster command that does wait anyway. (It that ever changes, we can handle refactor then.)

```suggestion
        } else if (c->cmd->proc == clusterCommand) {
```

**Review by zuiderkwast — COMMENTED** [review]
Very good cleanup. The PR description helped me a lot! Thanks.

If you decide to take my suggestions, don't forget to add signoff.

**Comment by PingXie** [comment]
> If you decide to take my suggestions, don't forget to add signoff.

I feel that comparing to the cmd structure point is cleaner/more readable but yes it is very subjective.

**Comment by zuiderkwast** [comment]
> > If you decide to take my suggestions, don't forget to add signoff.
> 
> I feel that comparing to the cmd structure point is cleaner/more readable but yes it is very subjective.

Pros and cons. It's perhaps cleaner, but it's more important IMO to be consistent with other code doing the same thing. (The alternative is to change all the existing command comparisons to your style.)

Avoiding the need to initialize and store these extra pointers is another point.

**Comment by PingXie** [comment]
>but it's more important IMO to be consistent with other code doing the same thing.

You got me on "consistency" :-). I will switch to `proc`.

**Review by zuiderkwast — COMMENTED** [review]
Looks good, but there's some crash in `test "XREAD: XADD + DEL + LPUSH should not awake client"`.

**Comment by PingXie** [comment]
This is not a crash but a behavior change caught by the test. The unification of the three blocked types,, incorrectly, allows an LPUSH operation to wake up a stream reader. 

On a closer look, I don't think there is an easy way to unify these 3 states, for two reasons:
1. each wait (list/zset/stream) can be initiated  by **multiple** commands. it is quite a hassle to track down all these commands;
2. stream can be waited on even when it is non-existent, which makes it impossible to infer the data type from `blockForKeys` implicitly

I reverted the last commit.

**Review by zuiderkwast — APPROVED** [review]

**Pull request merged**